File: web/src/engine/adaptation/adapter.ts
// Portage de Adaptation/Adapter.swift — moteur de règles d'adaptation ordonné.
import type { AthleteProfile } from '../domain/profile';
import type { BodyZone, InjuryRecord } from '../domain/profile';
import type { Equipment } from '../domain/equipment';
import { canPractice } from '../domain/equipment';
import type { CompletedActivity, DailyReadiness } from '../domain/activity';
import type { SessionIntent, Sport } from '../domain/types';
import type { PlannedSession } from '../domain/session';
import type { TrainingZones } from '../domain/zones';
import type { LoadPoint } from '../load/loadModel';
import { SessionGenerator } from '../sessionGeneration/sessionGenerator';
import { EquipmentSubstitution } from './equipmentSubstitution';
import { ReadinessEvaluator } from './readinessEvaluator';
import { isTestSession } from '../planning/testSessions';
import { addDays, isSameDay } from '../util/dates';

export type AdaptationEventKind =
  | 'eased'
  | 'rescheduled'
  | 'deload'
  | 'substituted'
  | 'injuryAdjusted'
  | 'alert'
  | 'recalibrated';

export interface AdaptationEvent {
  id: string;
  kind: AdaptationEventKind;
  date: Date;
  sessionID?: string;
  message: string;
}

let eventCounter = 0;
function makeEvent(kind: AdaptationEventKind, date: Date, message: string, sessionID?: string): AdaptationEvent {
  return { id: `evt-${++eventCounter}`, kind, date, sessionID, message };
}

export interface AdaptationContext {
  today: Date;
  upcoming: PlannedSession[];
  missed?: PlannedSession[];
  recentLoad?: LoadPoint[];
  readinessToday?: DailyReadiness;
  readinessHistory?: DailyReadiness[];
  injuries?: InjuryRecord[];
  equipment: Equipment;
  profile: AthleteProfile;
  zones: TrainingZones;
}

export interface AdaptationResult {
  plan: PlannedSession[];
  events: AdaptationEvent[];
}

export const medicalDisclaimer =
  'Ceci ne remplace pas un avis médical : si la douleur persiste, consulte un professionnel.';

export class Adapter {
  private generator: SessionGenerator;
  private substitution = new EquipmentSubstitution();
  private readiness = new ReadinessEvaluator();

  constructor(generator: SessionGenerator = new SessionGenerator()) {
    this.generator = generator;
  }

  adapt(ctx: AdaptationContext): AdaptationResult {
    // Les séances de TEST sont sanctuarisées : jamais rematérialisées ni allégées
    // (leur protocole doit rester intact pour calibrer les référentiels).
    const tests = ctx.upcoming.filter(isTestSession);
    let plan = ctx.upcoming.filter((s) => !isTestSession(s)).sort((a, b) => a.date.getTime() - b.date.getTime());
    let events: AdaptationEvent[] = [];

    ({ plan, events } = this.applyInjuries(plan, events, ctx));
    ({ plan, events } = this.applyEquipment(plan, events, ctx));
    ({ plan, events } = this.applyReadiness(plan, events, ctx));
    ({ plan, events } = this.applyLoadGuards(plan, events, ctx));
    ({ plan, events } = this.applyReschedule(plan, events, ctx));

    return { plan: [...plan, ...tests].sort((a, b) => a.date.getTime() - b.date.getTime()), events };
  }

  // 1. Blessures
  private injuryPolicy(zone: BodyZone): { forbidden: Set<Sport>; safe: Sport; hint: string } {
    switch (zone) {
      case 'knee': return { forbidden: new Set(['run', 'brick']), safe: 'swim', hint: 'Genou : natation avec pull buoy conservée, vélo allégé, course suspendue.' };
      case 'ankle':
      case 'foot':
      case 'calf': return { forbidden: new Set(['run', 'brick']), safe: 'swim', hint: 'Cheville/pied : impact évité, natation et vélo maintenus.' };
      case 'hamstring': return { forbidden: new Set(['run', 'brick']), safe: 'bike', hint: 'Ischios : pas de vitesse à pied, vélo souple et natation ok.' };
      case 'hip': return { forbidden: new Set(['run', 'brick']), safe: 'swim', hint: 'Hanche : on privilégie natation et vélo doux.' };
      case 'shoulder': return { forbidden: new Set(['swim']), safe: 'bike', hint: 'Épaule : natation suspendue, vélo et course maintenus.' };
      case 'lowerBack': return { forbidden: new Set(['brick']), safe: 'bike', hint: 'Bas du dos : intensité réduite, position vélo surveillée.' };
      case 'other': return { forbidden: new Set([]), safe: 'swim', hint: 'Zone sensible : prudence sur les séances intenses.' };
    }
  }

  private applyInjuries(plan: PlannedSession[], events: AdaptationEvent[], ctx: AdaptationContext): AdaptationResult {
    const out = [...plan];
    const ev = [...events];
    const active = (ctx.injuries ?? []).filter((i) => i.intensity >= 3);
    if (active.length === 0) return { plan: out, events: ev };

    for (const injury of active) {
      const policy = this.injuryPolicy(injury.zone);
      for (let i = 0; i < out.length; i++) {
        if (!policy.forbidden.has(out[i].sport)) continue;
        const safe = canPractice(ctx.equipment, policy.safe) ? policy.safe : 'strength';
        const repl = this.rematerialize(out[i], safe, 'recovery', ctx);
        repl.notes = `${policy.hint} ${medicalDisclaimer}`;
        ev.push(makeEvent('injuryAdjusted', out[i].date, `${policy.hint} ${medicalDisclaimer}`, out[i].id));
        out[i] = repl;
      }
    }
    return { plan: out, events: ev };
  }

  // 2. Matériel
  private applyEquipment(plan: PlannedSession[], events: AdaptationEvent[], ctx: AdaptationContext): AdaptationResult {
    const out = [...plan];
    const ev = [...events];
    for (let i = 0; i < out.length; i++) {
      const r = this.substitution.substitute(out[i], ctx.equipment);
      if (!r.changed) continue;
      const repl = this.rematerialize(out[i], r.session.sport, out[i].intent, ctx);
      repl.title = r.session.title;
      repl.notes = r.explanation;
      repl.estimatedLoad = r.session.estimatedLoad;
      out[i] = repl;
      ev.push(makeEvent('substituted', out[i].date, r.explanation, out[i].id));
    }
    return { plan: out, events: ev };
  }

  // 3. Récupération du jour
  private applyReadiness(plan: PlannedSession[], events: AdaptationEvent[], ctx: AdaptationContext): AdaptationResult {
    if (!ctx.readinessToday) return { plan, events };
    const a = this.readiness.assess(ctx.readinessToday, ctx.readinessHistory ?? []);
    if (a.level === 'good') return { plan, events };

    const out = [...plan];
    const ev = [...events];
    const todays = out.map((_, i) => i).filter((i) => isSameDay(out[i].date, ctx.today));
    if (todays.length === 0) return { plan: out, events: ev };
    const hardest = todays.reduce((best, i) => (out[i].estimatedLoad > out[best].estimatedLoad ? i : best), todays[0]);

    if (a.level === 'low') {
      out[hardest] = this.rematerialize(out[hardest], out[hardest].sport, 'recovery', ctx);
      ev.push(makeEvent('eased', out[hardest].date, `Récupération basse — on allège aujourd'hui pour mieux progresser demain. (${a.reasons[0] ?? ''})`, out[hardest].id));
    } else {
      const softer = this.downshift(out[hardest].intent);
      if (softer !== out[hardest].intent) {
        out[hardest] = this.rematerialize(out[hardest], out[hardest].sport, softer, ctx);
        ev.push(makeEvent('eased', out[hardest].date, "Forme moyenne : intensité réduite d'un cran aujourd'hui.", out[hardest].id));
      }
    }
    return { plan: out, events: ev };
  }

  private downshift(intent: SessionIntent): SessionIntent {
    switch (intent) {
      case 'vo2':
      case 'sprint': return 'threshold';
      case 'threshold': return 'tempo';
      case 'tempo': return 'endurance';
      default: return intent;
    }
  }

  // 4. Garde-fous de charge
  private applyLoadGuards(plan: PlannedSession[], events: AdaptationEvent[], ctx: AdaptationContext): AdaptationResult {
    const last = (ctx.recentLoad ?? []).at(-1);
    if (!last) return { plan, events };
    const out = [...plan];
    const ev = [...events];

    if (last.acwr > 1.5) {
      const horizon = addDays(ctx.today, 7);
      for (let i = 0; i < out.length; i++) {
        if (out[i].date <= horizon) {
          out[i] = { ...out[i], estimatedLoad: Math.round(out[i].estimatedLoad * 0.7), estimatedDuration: out[i].estimatedDuration * 0.8 };
        }
      }
      ev.push(makeEvent('deload', ctx.today, `Charge aiguë élevée (ACWR ${last.acwr.toFixed(2)}) : semaine allégée de 30 % pour prévenir la blessure.`));
    }

    if (last.tsb < -25) {
      ev.push(makeEvent('alert', ctx.today, `Fraîcheur très basse (TSB ${Math.trunc(last.tsb)}) : priorité au sommeil, on lève le pied 48 h.`));
    }
    return { plan: out, events: ev };
  }

  // 5. Rattrapage des séances manquées
  private applyReschedule(plan: PlannedSession[], events: AdaptationEvent[], ctx: AdaptationContext): AdaptationResult {
    const missed = ctx.missed ?? [];
    if (missed.length === 0) return { plan, events };
    const out = [...plan];
    const ev = [...events];
    const acwr = (ctx.recentLoad ?? []).at(-1)?.acwr ?? 0;

    for (const m of missed) {
      if (acwr > 1.2) {
        ev.push(makeEvent('alert', ctx.today, `Séance « ${m.title} » non rattrapée volontairement : on ne surcharge pas une semaine déjà lourde.`, m.id));
        continue;
      }
      let placed = false;
      for (let offset = 1; offset <= 7; offset++) {
        const day = addDays(ctx.today, offset);
        const occupied = out.some((s) => isSameDay(s.date, day));
        if (!occupied) {
          const moved = { ...m, date: day, estimatedLoad: Math.round(m.estimatedLoad * 0.9) };
          out.push(moved);
          ev.push(makeEvent('rescheduled', day, `Séance « ${m.title} » replacée sans surcharger la semaine.`, m.id));
          placed = true;
          break;
        }
      }
      if (!placed) {
        ev.push(makeEvent('alert', ctx.today, `Pas de créneau libre pour « ${m.title} » : on la laisse filer plutôt que d'entasser.`, m.id));
      }
    }
    return { plan: out, events: ev };
  }

  private rematerialize(s: PlannedSession, sport: Sport, intent: SessionIntent, ctx: AdaptationContext): PlannedSession {
    const nw = this.generator.generate(sport, intent, s.phase ?? 'base', s.date, ctx.zones, ctx.profile, ctx.equipment);
    nw.id = s.id;
    return nw;
  }

  /** Détecte les séances passées non réalisées (même jour + même sport). */
  static detectMissed(pastPlanned: PlannedSession[], completed: CompletedActivity[]): PlannedSession[] {
    return pastPlanned.filter(
      (planned) => !completed.some((act) => act.sport === planned.sport && isSameDay(act.start, planned.date)),
    );
  }
}

File: Packages/TriathlonEngine/Sources/TriathlonEngine/Adaptation/Adapter.swift
import Foundation

/// Événement d'adaptation : chaque ajustement automatique est tracé et expliqué
/// (affichable et, côté app, annulable).
public struct AdaptationEvent: Sendable, Equatable, Identifiable {
    public enum Kind: String, Sendable, Codable {
        case eased, rescheduled, deload, substituted, injuryAdjusted, alert, recalibrated
    }
    public var id: UUID
    public var kind: Kind
    public var date: Date
    public var sessionID: UUID?
    public var message: String

    public init(id: UUID = UUID(), kind: Kind, date: Date, sessionID: UUID? = nil, message: String) {
        self.id = id; self.kind = kind; self.date = date; self.sessionID = sessionID; self.message = message
    }
}

/// Contexte d'entrée du moteur d'adaptation.
public struct AdaptationContext: Sendable {
    public var today: Date
    public var upcoming: [PlannedSession]        // séances >= aujourd'hui, triées par date
    public var missed: [PlannedSession]          // séances passées non réalisées
    public var recentLoad: [LoadPoint]           // pour ACWR / TSB (le dernier = le plus récent)
    public var readinessToday: DailyReadiness?
    public var readinessHistory: [DailyReadiness]
    public var injuries: [InjuryRecord]
    public var equipment: Equipment
    public var profile: AthleteProfile
    public var zones: TrainingZones

    public init(
        today: Date,
        upcoming: [PlannedSession],
        missed: [PlannedSession] = [],
        recentLoad: [LoadPoint] = [],
        readinessToday: DailyReadiness? = nil,
        readinessHistory: [DailyReadiness] = [],
        injuries: [InjuryRecord] = [],
        equipment: Equipment,
        profile: AthleteProfile,
        zones: TrainingZones
    ) {
        self.today = today
        self.upcoming = upcoming
        self.missed = missed
        self.recentLoad = recentLoad
        self.readinessToday = readinessToday
        self.readinessHistory = readinessHistory
        self.injuries = injuries
        self.equipment = equipment
        self.profile = profile
        self.zones = zones
    }
}

public struct AdaptationResult: Sendable {
    public var plan: [PlannedSession]
    public var events: [AdaptationEvent]
}

/// Rappel médical systématique, jamais absent des ajustements liés à la douleur.
public let medicalDisclaimer = "Ceci ne remplace pas un avis médical : si la douleur persiste, consulte un professionnel."

/// Moteur de règles d'adaptation. Applique, dans l'ordre : sécurité (blessures),
/// disponibilité matériel, récupération du jour, garde-fous de charge, rattrapage.
public struct Adapter: Sendable {
    private let generator: SessionGenerator
    private let substitution = EquipmentSubstitution()
    private let readiness = ReadinessEvaluator()

    public init(generator: SessionGenerator = SessionGenerator()) {
        self.generator = generator
    }

    public func adapt(_ ctx: AdaptationContext) -> AdaptationResult {
        var plan = ctx.upcoming.sorted { $0.date < $1.date }
        var events: [AdaptationEvent] = []

        // 1) Blessures — priorité sécurité.
        (plan, events) = applyInjuries(plan, events: events, ctx: ctx)

        // 2) Disponibilité matériel.
        (plan, events) = applyEquipment(plan, events: events, ctx: ctx)

        // 3) Récupération du jour → allègement de la séance du jour si nécessaire.
        (plan, events) = applyReadiness(plan, events: events, ctx: ctx)

        // 4) Garde-fous de charge (ACWR / TSB).
        (plan, events) = applyLoadGuards(plan, events: events, ctx: ctx)

        // 5) Rattrapage intelligent des séances manquées.
        (plan, events) = applyReschedule(plan, events: events, ctx: ctx)

        return AdaptationResult(plan: plan.sorted { $0.date < $1.date }, events: events)
    }

    // MARK: - 1. Blessures

    /// Sports interdits et alternative sûre selon la zone blessée.
    private func injuryPolicy(_ zone: InjuryRecord.BodyZone) -> (forbidden: Set<Sport>, safe: Sport, hint: String) {
        switch zone {
        case .knee: return ([.run, .brick], .swim, "Genou : natation avec pull buoy conservée, vélo allégé, course suspendue.")
        case .ankle, .foot, .calf: return ([.run, .brick], .swim, "Cheville/pied : impact évité, natation et vélo maintenus.")
        case .hamstring: return ([.run, .brick], .bike, "Ischios : pas de vitesse à pied, vélo souple et natation ok.")
        case .hip: return ([.run, .brick], .swim, "Hanche : on privilégie natation et vélo doux.")
        case .shoulder: return ([.swim], .bike, "Épaule : natation suspendue, vélo et course maintenus.")
        case .lowerBack: return ([.brick], .bike, "Bas du dos : intensité réduite, position vélo surveillée.")
        case .other: return ([], .swim, "Zone sensible : prudence sur les séances intenses.")
        }
    }

    private func applyInjuries(_ plan: [PlannedSession], events: [AdaptationEvent], ctx: AdaptationContext) -> ([PlannedSession], [AdaptationEvent]) {
        var out = plan, ev = events
        let active = ctx.injuries.filter { $0.intensity >= 3 }
        guard !active.isEmpty else { return (out, ev) }

        for injury in active {
            let policy = injuryPolicy(injury.zone)
            for i in out.indices where policy.forbidden.contains(out[i].sport) {
                let safe = ctx.equipment.canPractice(policy.safe) ? policy.safe : .strength
                var repl = rematerialize(out[i], sport: safe, intent: .recovery, ctx: ctx)
                repl.notes = "\(policy.hint) \(medicalDisclaimer)"
                ev.append(AdaptationEvent(kind: .injuryAdjusted, date: out[i].date, sessionID: out[i].id,
                                          message: policy.hint + " " + medicalDisclaimer))
                out[i] = repl
            }
        }
        return (out, ev)
    }

    // MARK: - 2. Matériel

    private func applyEquipment(_ plan: [PlannedSession], events: [AdaptationEvent], ctx: AdaptationContext) -> ([PlannedSession], [AdaptationEvent]) {
        var out = plan, ev = events
        for i in out.indices {
            let r = substitution.substitute(out[i], equipment: ctx.equipment)
            guard r.changed else { continue }
            // Re-matérialise avec le générateur pour un contenu cohérent du nouveau sport.
            var repl = rematerialize(out[i], sport: r.session.sport, intent: out[i].intent, ctx: ctx)
            repl.title = r.session.title
            repl.notes = r.explanation
            repl.estimatedLoad = r.session.estimatedLoad
            out[i] = repl
            ev.append(AdaptationEvent(kind: .substituted, date: out[i].date, sessionID: out[i].id, message: r.explanation))
        }
        return (out, ev)
    }

    // MARK: - 3. Récupération du jour

    private func applyReadiness(_ plan: [PlannedSession], events: [AdaptationEvent], ctx: AdaptationContext) -> ([PlannedSession], [AdaptationEvent]) {
        guard let today = ctx.readinessToday else { return (plan, events) }
        let a = readiness.assess(today: today, history: ctx.readinessHistory)
        guard a.level != .good else { return (plan, events) }

        var out = plan, ev = events
        let cal = Calendar(identifier: .gregorian)
        // Séance la plus dure du jour.
        let todays = out.indices.filter { cal.isDate(out[$0].date, inSameDayAs: ctx.today) }
        guard let hardest = todays.max(by: { out[$0].estimatedLoad < out[$1].estimatedLoad }) else { return (out, ev) }

        if a.level == .low {
            let repl = rematerialize(out[hardest], sport: out[hardest].sport, intent: .recovery, ctx: ctx)
            out[hardest] = repl
            ev.append(AdaptationEvent(kind: .eased, date: out[hardest].date, sessionID: out[hardest].id,
                                      message: "Récupération basse — on allège aujourd'hui pour mieux progresser demain. (\(a.reasons.first ?? ""))"))
        } else {
            // Modérée : on baisse l'intensité d'un cran sans changer le sport.
            let softer = downshift(out[hardest].intent)
            if softer != out[hardest].intent {
                let repl = rematerialize(out[hardest], sport: out[hardest].sport, intent: softer, ctx: ctx)
                out[hardest] = repl
                ev.append(AdaptationEvent(kind: .eased, date: out[hardest].date, sessionID: out[hardest].id,
                                          message: "Forme moyenne : intensité réduite d'un cran aujourd'hui."))
            }
        }
        return (out, ev)
    }

    private func downshift(_ intent: SessionIntent) -> SessionIntent {
        switch intent {
        case .vo2, .sprint: return .threshold
        case .threshold: return .tempo
        case .tempo: return .endurance
        default: return intent
        }
    }

    // MARK: - 4. Garde-fous de charge

    private func applyLoadGuards(_ plan: [PlannedSession], events: [AdaptationEvent], ctx: AdaptationContext) -> ([PlannedSession], [AdaptationEvent]) {
        guard let last = ctx.recentLoad.last else { return (plan, events) }
        var out = plan, ev = events
        let cal = Calendar(identifier: .gregorian)

        // ACWR trop haut → décharge anticipée des 7 prochains jours (−30 %).
        if last.acwr > 1.5 {
            let horizon = cal.date(byAdding: .day, value: 7, to: ctx.today)!
            for i in out.indices where out[i].date <= horizon {
                out[i].estimatedLoad = (out[i].estimatedLoad * 0.7).rounded()
                out[i].estimatedDuration *= 0.8
            }
            ev.append(AdaptationEvent(kind: .deload, date: ctx.today,
                                      message: "Charge aiguë élevée (ACWR \(String(format: "%.2f", last.acwr))) : semaine allégée de 30 % pour prévenir la blessure."))
        }

        // TSB très négatif → micro-décharge + priorité sommeil.
        if last.tsb < -25 {
            ev.append(AdaptationEvent(kind: .alert, date: ctx.today,
                                      message: "Fraîcheur très basse (TSB \(Int(last.tsb))) : priorité au sommeil, on lève le pied 48 h."))
        }
        return (out, ev)
    }

    // MARK: - 5. Rattrapage des séances manquées

    private func applyReschedule(_ plan: [PlannedSession], events: [AdaptationEvent], ctx: AdaptationContext) -> ([PlannedSession], [AdaptationEvent]) {
        guard !ctx.missed.isEmpty else { return (plan, events) }
        var out = plan, ev = events
        let cal = Calendar(identifier: .gregorian)
        let acwr = ctx.recentLoad.last?.acwr ?? 0

        for missed in ctx.missed {
            // On ne rattrape pas si la charge est déjà trop haute.
            if acwr > 1.2 {
                ev.append(AdaptationEvent(kind: .alert, date: ctx.today, sessionID: missed.id,
                                          message: "Séance « \(missed.title) » non rattrapée volontairement : on ne surcharge pas une semaine déjà lourde."))
                continue
            }
            // Cherche un créneau libre dans les 7 jours.
            var placed = false
            for offset in 1...7 {
                let day = cal.date(byAdding: .day, value: offset, to: ctx.today)!
                let occupied = out.contains { cal.isDate($0.date, inSameDayAs: day) }
                if !occupied {
                    var moved = missed
                    moved.date = day
                    moved.estimatedLoad = (missed.estimatedLoad * 0.9).rounded()
                    out.append(moved)
                    ev.append(AdaptationEvent(kind: .rescheduled, date: day, sessionID: missed.id,
                                              message: "Séance « \(missed.title) » replacée sans surcharger la semaine."))
                    placed = true
                    break
                }
            }
            if !placed {
                ev.append(AdaptationEvent(kind: .alert, date: ctx.today, sessionID: missed.id,
                                          message: "Pas de créneau libre pour « \(missed.title) » : on la laisse filer plutôt que d'entasser."))
            }
        }
        return (out, ev)
    }

    // MARK: - Utilitaires

    /// Reconstruit une séance (id & date conservés) avec un nouveau sport/intention.
    private func rematerialize(_ s: PlannedSession, sport: Sport, intent: SessionIntent, ctx: AdaptationContext) -> PlannedSession {
        var new = generator.generate(sport: sport, intent: intent, phase: s.phase ?? .base,
                                     date: s.date, zones: ctx.zones, profile: ctx.profile, equipment: ctx.equipment)
        new.id = s.id
        return new
    }

    // MARK: - Détection des séances manquées (utilitaire pur)

    /// Compare les séances passées planifiées aux activités réalisées (même jour + même sport).
    public static func detectMissed(pastPlanned: [PlannedSession], completed: [CompletedActivity],
                                    calendar: Calendar = .init(identifier: .gregorian)) -> [PlannedSession] {
        pastPlanned.filter { planned in
            !completed.contains { act in
                act.sport == planned.sport && calendar.isDate(act.start, inSameDayAs: planned.date)
            }
        }
    }
}
